Lib/fontTools/varLib/__init__.py

```diff
@@ -208,6 +212,62 @@ def _add_avar(font, axes):
         assert +1.0 not in curve or curve[+1.0] == +1.0
         # curve.update({-1.0: -1.0, 0.0: 0.0, 1.0: 1.0})
 
+    if mappings:
+        interesting = True
+
+        hiddenAxes = [axis for axis in axes.values() if axis.hidden]
+
+        inputLocations = [
+            {
+                axes[name].tag: models.normalizeValue(v, vals_triples[axes[name].tag])
+                for name, v in mapping.inputLocation.items()
+            }
+            for mapping in mappings
+        ]
+        outputLocations = [
+            {
+                axes[name].tag: models.normalizeValue(v, vals_triples[axes[name].tag])
+                for name, v in mapping.outputLocation.items()
+            }
+            for mapping in mappings
+        ]
+
+        # If base-master is missing, insert it at zero location.
+        if not any(all(v == 0 for k, v in loc.items()) for loc in inputLocations):
+            inputLocations.insert(0, {})
+            outputLocations.insert(0, {})
+
+        model = models.VariationModel(inputLocations, axisTags)
+        storeBuilder = varStore.OnlineVarStoreBuilder(axisTags)
+        storeBuilder.setModel(model)
+        varIdxes = {}
+        for t in axisTags:
+            masterValues = []
+            for vo, vi in zip(outputLocations, inputLocations):
+                if t not in vo:
+                    masterValues.append(0)
+                    continue
+                if t not in vi and t not in hiddenAxes:
+                    log.warning(
+                        "No input location specified for non-hidden axis '%s' in axis mapping %s",
```

Comment:
maybe the warning could mention what's the consequence of not specifying an input location for a non-hidden axis

Reply:
Done. The 0 means axis default value.